From 416a14ed6d5fe4e48f19a7f4872714cf43b1933c Mon Sep 17 00:00:00 2001
From: John Zhuge <jzhuge@cloudera.com>
Date: Thu, 11 Aug 2016 14:22:56 -0700
Subject: [PATCH 1741/2748] HDFS-10245. Fix the findbugs warnings in
 branch-2.7. Contributed by Brahma Reddy Battula.

(cherry picked from commit 076215024bd3811a2b05e26c501f8b1ec27904ce)

Change-Id: I99eb436b2bce63bbef00b399621ade542b62149b
---
 .../dev-support/findbugsExcludeFile.xml            |   11 +++++++++++
 .../org/apache/hadoop/hdfs/DFSOutputStream.java    |    4 ++--
 2 files changed, 13 insertions(+), 2 deletions(-)

diff --git a/hadoop-hdfs-project/hadoop-hdfs/dev-support/findbugsExcludeFile.xml b/hadoop-hdfs-project/hadoop-hdfs/dev-support/findbugsExcludeFile.xml
index f7e93b3..fa511fb 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/dev-support/findbugsExcludeFile.xml
+++ b/hadoop-hdfs-project/hadoop-hdfs/dev-support/findbugsExcludeFile.xml
@@ -79,6 +79,17 @@
        <Bug pattern="REC_CATCH_EXCEPTION" />
      </Match>
 
+    <!--
+     we don't need to synchronize out in newStreamForAppend method because
+     out.streamer cannot be touched by other threads. It is false-positive.
+    -->
+    <Match>
+        <Class name="org.apache.hadoop.hdfs.DFSOutputStream" />
+        <Method name="newStreamForAppend" />
+        <Field name="streamer" />
+        <Bug pattern="IS2_INCONSISTENT_SYNC" />
+    </Match>
+
      <!--
       lastAppliedTxid is carefully unsynchronized in the BackupNode in a couple spots.
       See the comments in BackupImage for justification.
diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/DFSOutputStream.java b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/DFSOutputStream.java
index 11e53c6..2c80bef 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/DFSOutputStream.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/DFSOutputStream.java
@@ -2662,7 +2662,7 @@ public void setDropBehind(Boolean dropBehind) throws IOException {
   }
 
   @VisibleForTesting
-  ExtendedBlock getBlock() {
+  synchronized ExtendedBlock getBlock() {
     return streamer.getBlock();
   }
 
@@ -2680,7 +2680,7 @@ public long getFileId() {
    * @return The times have retried to recover pipeline, for the same packet.
    */
   @VisibleForTesting
-  int getPipelineRecoveryCount() {
+  synchronized int getPipelineRecoveryCount() {
     return streamer.pipelineRecoveryCount;
   }
 }
-- 
1.7.9.5

